cmreactor/cmreactor.sh

```diff
@@ -0,0 +1,483 @@
+    local branch
+    local extractedDir
+    local modulePath
+    local metaFile
+
+    config="$1"
+    module="$2"
+    tempBase="$3"
+
+    modulePath="$(pathInProject "${module}")"
+    url="$(readModuleConfigValue "${configFile}" "${module}" "url")"
+    if [[ "$url" == "" ]]; then
+        return 1
+    elif [[ "$url" == "." ]]; then
+        url="$(pwd)"
+    fi
+
+    if [[ -d "${url}/.git" ]]; then
+        origin="$(git --no-pager -C "${url}" remote get-url origin)"
+    else
+        origin="$url"
+    fi
+
+    extractedDir="${tempBase}/${module}"
+    [[ -d "${extractedDir}" ]] && rm -rf "${extractedDir}"
+
+    git clone "${url}" "${extractedDir}"
+    branch="$(readModuleConfigValue "${configFile}" "${module}" "branch")"
+    if [[ "$branch" != "" ]]; then
+        git --no-pager -C "${extractedDir}" checkout "$branch"
+    fi
+    git --no-pager -C "${extractedDir}" remote set-url origin "${origin}"
+
+    [[ -d "${modulePath}" ]] && rm -rf "${modulePath}"
+    mv "${extractedDir}" "${modulePath}"
+
+    metaFile="${modulePath}/cmreactor-meta.gitconfig"
+    gitp config -f "${metaFile}" "cmreactor.meta.${module}.type" "git"
+    gitp config -f "${metaFile}" "cmreactor.meta.${module}.url" "${url}"
+    gitp config -f "${metaFile}" "cmreactor.meta.${module}.origin" "${origin}"
+    gitp config -f "${metaFile}" "cmreactor.meta.${module}.branch" "$(git --no-pager -C "${modulePath}" rev-parse --abbrev-ref HEAD)"
+    gitp config -f "${metaFile}" "cmreactor.meta.${module}.commit" "$(git --no-pager -C "${modulePath}" rev-parse HEAD)"
+    gitp config -f "${metaFile}" "cmreactor.meta.${module}.oneline" "$(git --no-pager -C "${modulePath}" show -s --oneline)"
+}
+
+recordReactorGitMeta() {
+    local metaFile
+    metaFile="$(pathInProject "cmreactor-meta.gitconfig")"
+
+    gitp config -f "${metaFile}" "cmreactor.reactor-meta.origin" "$(gitc remote get-url origin)"
+    gitp config -f "${metaFile}" "cmreactor.reactor-meta.branch" "$(gitc rev-parse --abbrev-ref HEAD)"
+    gitp config -f "${metaFile}" "cmreactor.reactor-meta.commit" "$(gitc rev-parse HEAD)"
+    gitp config -f "${metaFile}" "cmreactor.reactor-meta.oneline" "$(gitc show -s --oneline)"
+}
+
+syncMavenRepositoryFolders() {
+    find "${projectPath}" -name maven_repository -type d -mindepth 2 \
+        -exec rsync -a '{}'/ "${projectPath}/maven_repository/" \;
+}
+### END CMREACTOR FUNCTIONS ###
+
+if [[ "${projectPath}" == "/" ]]; then
+  echo "$(basename "${BASH_SOURCE[0]}") prevented root path removal" >&2
+  exit 1
+fi
+
+[[ -d "${projectPath}" ]] && rm -rf "${projectPath}"
+mkdir -p "${projectPath}"
+
+declare -a moduleNames
+moduleNames=($(readModulesFromConfig "${configFile}"))
+
+reactorPomInit "${moduleNames[@]}"
+recordReactorGitMeta
+
+if [[ "${moduleNames[*]}" != "" ]]; then
+    tempBase="$(pathInSource target/cmreactor-tmp)"
+    mkdir -p "${tempBase}"
+    for moduleName in "${moduleNames[@]}"; do
+        if [[ "${moduleName}" =~ [^A-Za-z0-9_.-] ]]; then
```

Comment:
If someone were to use a `moduleName` that matches a module of the current project, would this cause that module to be unintentionally deleted?

Reply:
No. But I mistakenly allowed dot (and dot-dot), which I have now corrected.